addrman: Make consistency checks a runtime option

jnewbery opened the pull request:
CAddrMan has internal consistency checks. Currently, these are only run when the program is compiled with the  `DEBUG_ADDRMAN` option. This option is not enabled on any of our CI builds, and it's likely that no-one is running them at all.

This PR makes consistency checks a (hidden) runtime option that can be enabled with `-checkaddrman`, where `-checkaddrman=n` will result in the consistency checks running every n operations (similar to `-checkmempool=n`). We set the ratio to 1/100 for our unit tests, and leave it disabled by default for all networks. Additionally, a consistency check failure now asserts, rather than logging and continuing. This matches the behavior of CTxMemPool and TxRequestTracker, where a failed consistency check asserts.

jnewbery committed:
[addrman] Make addrman consistency checks a runtime option

Currently addrman consistency checks are a compile time option, and are not
enabled in our CI. It's unlikely anyone is running these consistency checks.

Make them a runtime option instead, where users can enable addrman
consistency checks every n operations (similar to mempool tests). Update
the addrman unit tests to do internal consistency checks every 100
operations (checking on every operations causes the test runtime to
increase by several seconds).

Also assert on a failed addrman consistency check to terminate program
execution.

Empact commented on src/test/util/setup_common.cpp:
```diff
@@ -172,7 +172,9 @@ TestingSetup::TestingSetup(const std::string& chainName, const std::vector<const
-    m_node.peerman = MakeUnique<PeerManager>(chainparams, *m_node.connman, m_node.banman.get(), *m_node.scheduler, *m_node.chainman, *m_node.mempool);
+    m_node.peerman = MakeUnique<PeerManager>(chainparams, *m_node.connman, *m_node.addrman,
+                                             m_node.banman.get(), *m_node.scheduler, *m_node.chainman,
+                                             *m_node.mempool);
```
nit: other cases line break after banman

jnewbery replied:
I don't it matters where the line break is, as long as the line isn't too long.

practicalswift commented:
Concept ACK

Should we run with `-checkaddrman` in at least one of the Traivs jobs?

jnewbery commented:
> Should we run with -checkaddrman in at least one of the Traivs jobs?

`-checkaddrman` is enabled by default for regtest, so is on for all Travis jobs (although our functional tests probably don't exercise addrman very much)

jnewbery commented:
The sanity check fails on one of the unit tests. I've converted this PR to draft status while I work out why it's failing.

practicalswift commented:
> `-checkaddrman` is enabled by default for regtest, so is on for all Travis jobs (although our functional tests probably don't exercise addrman very much)

Oh, I missed the "by default on regtest" in the PR description. Sorry about that :)

Defaulting to sanity checking in regtest makes perfect sense! Even more Concept ACK :)

jnewbery commented:
I've fixed the failing unit tests.

jnewbery commented:
This makes some of the functional tests run very slowly. I'll do some comparisons and maybe disable consistency checks on the slowest tests.

jnewbery commented:
Marking as draft until #20228 is merged.

ajtowns commented:
This seems to make rpc_net and p2p_getaddr_caching tests run extremely slowly:

    rpc_net.py                                       | ✓ Passed  | 1372 s
    p2p_getaddr_caching.py                           | ✓ Passed  | 1367 s

I don't think there's any need for this to be based on #20228 -- passing a "check_addrman" from init to CConnman seems to work fine, see https://github.com/ajtowns/bitcoin/commits/202101-addrman-check eg.

More generally, I'm not sure that these consistency checks are that useful to enable at runtime/CI rather than just using them as an aid when making changes to addrman internals.

jnewbery commented:
> This seems to make rpc_net and p2p_getaddr_caching tests run extremely slowly.

Yes, see my comment at https://github.com/bitcoin/bitcoin/pull/20233#issuecomment-716160060. I was planning on addressing that after 20228 was merged, probably by disabling the checks for those tests, or perhaps by changing the consistency check configuration to be a ratio, like the mempool consistency checks:

https://github.com/bitcoin/bitcoin/blob/68196a891056f98c1df0debacf09fb2ea4682a43/src/txmempool.h#L491

> I don't think there's any need for this to be based on #20228 -- passing a "check_addrman" from init to CConnman seems to work fine, see ajtowns/bitcoin@202101-addrman-check (commits) eg.

Sure, we could continue to pass function calls and (now) ctor parameters through CConnman, but I don't see why that is desirable. Connman _uses_ addrman, but there's no reason for it to _have_ an addrman. Addrman is also used by peerman and the RPCs. Connman's resposibility is to open and maintain connections. Storing and maintaining our map of accessible nodes on the network is a separate  responsibility.

There was also some discussion here: http://www.erisian.com.au/bitcoin-core-dev/log-2020-12-16.html#l-585 about being able to manage addrman more directly. Again, in that case it'd be easier if RPC directly held a handle to the addrman instead of passing messages through connman.

> More generally, I'm not sure that these consistency checks are that useful to enable at runtime/CI rather than just using them as an aid when making changes to addrman internals.

I think somewhere that they're definitely useful would be in fuzz testing, to ensure that there's no way to violate addrman's invariants. See the usage of txrequest's sanity check here:

https://github.com/bitcoin/bitcoin/blob/68196a891056f98c1df0debacf09fb2ea4682a43/src/test/fuzz/txrequest.cpp#L373

which is called after every loop in the fuzz tester.

ajtowns commented:
> Connman uses addrman, but there's no reason for it to have an addrman.

There's no point having addrman without a connman outside of unit tests (in which case the test can create an addrman directly), and connman uses addrman. That's enough reason for connman to be the thing that owns it, and for anything that wants it to access it via connman. I don't think there's anything you could do with `node.addrman` that you couldn't do equally well with `node.connman.GetAddrman()`.

But more importantly, it's easy to separate the two questions (should connman own the addrman instance and should debug_addrman be used more often) and judge them independently, so we should.

> I think somewhere that they're definitely useful would be in fuzz testing, [...]

Yeah; I agree in principle. But while I think txrequest's checks are reasonably optimised to be run frequently with modest data sets, I think addrman's is very heavy weight and only really works for pretty small ones. Maybe it would make sense to have the automatic internal checks (`Good()` surrounds `Good_()` by invocations to `Check()`) only happen via compile-time changes so you can use them when debugging, but still have `Check_()` compiled unconditionally so that the fuzzer and unit tests call `Check()` themselves.

Any idea how much this PR currently slows down the fuzzer? Doing the checks irregularly might be enough either way though -- perhaps especially if how irregular it is adjust dynamically (100% for the first x checks, 50% for the next x checks, etc?)

Hmm, actually, at least as it stands, perhaps addrman checks should be a regtest-only option, rather than a regtest-by-default option? Aren't those checks always too heavyweight to run when addrman is populated from mainnet/testnet?

jnewbery commented:
> But more importantly, it's easy to separate the two questions (should connman own the addrman instance and should debug_addrman be used more often) and judge them independently, so we should.

The aim of this PR is to add the ability to run internal consistency checks in addrman. If that can be done without adding more unnecessary coupling between components, then we should do that. There's no good reason that connman should be forwarding constructor parameters through to addrman, so I don't want to add that.

This design of both connman and peerman holding a handle to another component is exactly the same as we use for banman, with the added bonus that banman is internally an optional component (although there's no exposed configuration option to disable it). That design ensures very clean separation between components with no unexpected dependencies creeping in, and no boilerplate forwarding code in connman.

Changing the constuctor parameters is changing the public interface of addrman, so there are two choices:

1. keep it in connman and update connman's ctor interface, adding yet more coupling between components
2. split out addrman and then add the ctor parameter

(2) is more work now, but it's a net removal of boilerplate code (-24 in #20228), and is a better design for the long term.

jnewbery commented:
Rebased on master. I'm leaving this as draft since I still need to update it to disable addrman consistency checks for certain tests.

vasild commented:
Concept ACK

jnewbery committed:
[tests] Make deterministic addrman use nKey = 1

addrman_tests fail when consistency checks are enabled, since the tests
set the deterministic test addrman's nKey value to zero, which is an
invalid value. Change this so that deterministic addrman's nKey value is
set to 1.

This requires updating a few tests that are using magic values derived
from nKey being set to 0.

jnewbery commented:
Addrman corruption is getting more attention recently, so I've rebased on master, with some small changes:

- the option is now an integer, where setting `-checkaddrman=n` will cause the consistency checks to be run every `n` operations (similar to `-checkmempool=n`).
- the consistency check ratio is set to 100 for unit tests. On my machine, running the addrman_tests took: 
  - with no consistency checks: 0.65s
  - with consistency check ratio set to 100: 0.71s
  - with consistency check ratio set to 1: 7.75s

  Running consistency checks on 1% of operations seems to give good increased coverage without impacting test runtime too much
- consistency checks are disabled by default for all networks, including regtest. That can be changed in a follow-up PR if people think it should be enabled by default for regtest.

jonatack reviewed (commented):
Concept ACK to facilitate using this check, and as recompiling after defining DEBUG_ADDRMAN is very slow. Perhaps use uint32 rather than int32.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1160,7 +1161,8 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
     assert(!node.addrman);
-    node.addrman = std::make_unique<CAddrMan>();
+    int32_t check_addrman = std::min<int32_t>(std::max<int32_t>(args.GetArg("-checkaddrman", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), 0), 1000000);
```
nit: https://en.cppreference.com/w/cpp/algorithm/clamp

jnewbery replied:
Nice. Done.

MarcoFalke commented on src/addrman.h:
```diff
@@ -723,22 +729,19 @@ class CAddrMan
+        int err = Check_();
         if (err) {
             LogPrintf("ADDRMAN CONSISTENCY CHECK FAILED!!! err=%i\n", err);
+            assert(false);
```
Nice. Verbosely complaining to developers about bugs makes them less likely to get missed.

MarcoFalke replied:
Unrelated improvement: Maybe just simplify the `if foo return -xx;` with `assert(!foo);` in the cpp file?

vasild replied:
```suggestion
            std::abort();
```

jnewbery replied:
@MarcoFalke I left the cpp file unchanged in case we want to later add some kind of recovery code instead of simply asserting on failure. Changing the `if foo return -xx;` to `assert(!foo)` would involve making changes in more places.

@vasild - What's the advantage of using `std::abort`? I see lots more examples of `assert(false)` in the code than `std::abort()`.

vasild replied:
`assert()` calls `abort()`, so it is the same thing in practice (modulo the "strange" message `assertion failed: false`). Feel free to ignore or use `assert(0)` or `assert(nullptr)`.

ajtowns replied:
`assert(false && "addrman consistency check failed");` perhaps? Or `assert(err == 0)` or `Assume(err == 0)`

MarcoFalke replied:
#22500 ;)

jnewbery replied:
Thanks @MarcoFalke. I've added the two commits from that PR into this branch.

MarcoFalke commented:
review ACK a343f9079c30e1dce6f13852ff84bab7dab3fcde

jamesob commented:
Concept ACK, seems like an obvious win. Will review soon.

vasild reviewed (approved):
ACK a343f9079c30e1dce6f13852ff84bab7dab3fcde

This results in about 8% slowdown of `./src/test/test_bitcoin` on my machine (approx 6m10s -> 6m45s).

jnewbery commented:
> This results in about 8% slowdown of ./src/test/test_bitcoin on my machine (approx 6m10s -> 6m45s).

Hmmm, that's rather worrying. I'd expect the only test for this to have a material impact on would be the addrman tests, which take less than a second to run. Perhaps the calls to `GetRand()` are expensive enough that even with consistency checks disabled, this has a negative effect on performance.

MarcoFalke commented:
`GetRand` seems unneeded wasteful anyway (it calls `SeedFast` on every call). Have you seen https://github.com/bitcoin/bitcoin/pull/20233#discussion_r672437419 ?

jnewbery commented:
> GetRand seems unneeded wasteful anyway (it calls SeedFast on every call). Have you seen #20233 (comment) ?

I see it. I'll try this with `insecure_rand` tomorrow. Thanks for the tip!

vasild commented on src/addrman.cpp:
```diff
@@ -460,11 +460,13 @@ CAddrInfo CAddrMan::Select_(bool newOnly)
     AssertLockHeld(cs);
 
+    if (m_consistency_check == 0) return 0;
+    if (GetRand(m_consistency_check) >= 1) return 0;
```
Or set `g_mock_deterministic_tests` to `true` for all tests :eyes: :bomb:

jnewbery replied:
> At least here we could use insecure_rand, which should be deterministic if initialized deterministically.

This has the unfortunate effect that the checks are run on the same operations for every run of the unit tests. Having them picked at random means that eventually they're run at every operation in the unit tests. Taking entropy from insecure_rand at different points would also invalidate the tests that have hard-coded values based on the determinstic insecure_rand.

jnewbery replied:
Leaving this as is for now. The `GetRand()` call doesn't seem to result in a performance degradation from my tests: https://github.com/bitcoin/bitcoin/pull/20233#issuecomment-883284166.

vasild commented on src/addrman.h:
```diff
@@ -723,22 +729,19 @@ class CAddrMan
     CAddrInfo SelectTriedCollision_() EXCLUSIVE_LOCKS_REQUIRED(cs);
 
     //! Consistency check
-    void Check()
```
Would it be a good idea to
1. do the check unconditionally at the end of `Unserialize()` (as if `m_consistency_check == 1`) and
2. if it fails, to treat it as disk corruption - throw `std::ios_base::failure` instead of `abort()`?

jnewbery replied:
Seems reasonable. This can be done as a follow-up if desired.

jnewbery commented:
> This results in about 8% slowdown of ./src/test/test_bitcoin on my machine (approx 6m10s -> 6m45s).

@vasild how many samples did you take? There's some variance in the time taken to run the entire test suite, so you'll need to run it several times to ensure that this isn't just noise. Here are my results:

Returning immediately from `_Check()`:

```
→ time `for i in {1..5}; do ./src/test/test_bitcoin; done`

[...]

real	4m1.543s
user	2m28.929s
sys	0m25.260s
```

And calling `GetRand(m_consistency_check)` in `_Check()`:

```
→ time `for i in {1..5}; do ./src/test/test_bitcoin; done`

[...]

real	4m0.947s
user	2m33.306s
sys	0m21.416s
```

So for me, there's essentially no difference.

jnewbery commented on src/addrman.h:
```diff
@@ -482,7 +485,7 @@ class CAddrMan
     }
 
-    CAddrMan()
+    CAddrMan(int32_t addrman_check=DEFAULT_ADDRMAN_CONSISTENCY_CHECKS) : m_consistency_check(addrman_check)
```
done

jnewbery commented on src/addrman.h:
```diff
@@ -723,22 +729,19 @@ class CAddrMan
-    void Check()
-        EXCLUSIVE_LOCKS_REQUIRED(cs)
+    void Check() EXCLUSIVE_LOCKS_REQUIRED(cs)
     {
-#ifdef DEBUG_ADDRMAN
         AssertLockHeld(cs);
-        const int err = Check_();
+
+        int err = Check_();
```
added back

jnewbery commented on src/init.cpp:
```diff
@@ -502,6 +502,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-checkblocks=<n>", strprintf("How many blocks to check at startup (default: %u, 0 = all)", DEFAULT_CHECKBLOCKS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-checklevel=<n>", strprintf("How thorough the block verification of -checkblocks is: %s (0-4, default: %u)", Join(CHECKLEVEL_DOC, ", "), DEFAULT_CHECKLEVEL), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-checkblockindex", strprintf("Do a consistency check for the block tree, chainstate, and other validation data structures occasionally. (default: %u, regtest: %u)", defaultChainParams->DefaultConsistencyChecks(), regtestChainParams->DefaultConsistencyChecks()), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-checkaddrman", strprintf("Do consistency checks on the addrman before and after every <n> operations (default: %u)", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
done

jnewbery commented:
Added the two commits from PR #22500, to resolve the various suggestions in this thread: https://github.com/bitcoin/bitcoin/pull/20233#discussion_r672204994.

jnewbery commented:
@MarcoFalke has just told me that he no longer believes that #22500 is the right approach, so I've removed those two commits again.

lsilva01 reviewed (approved):
Code Review ACK and Tested ACK https://github.com/bitcoin/bitcoin/pull/20233/commits/b251f68ba7069fdb11cb66c0b6b8acdcce667fc9 on Ubuntu 20.04 using the runtime argument `-checkaddrman=100` on mainnet.
Unit tests (including `--run_test=addrman_tests`) have run successfully.

sipa commented on src/addrman.cpp:
```diff
@@ -459,11 +459,14 @@ CAddrInfo CAddrMan::Select_(bool newOnly)
 
+    // Run consistency checks 1 in m_consistency_check_ratio times if enabled
+    if (m_consistency_check_ratio == 0) return 0;
+    if (GetRand(m_consistency_check_ratio) >= 1) return 0;
```
Perhaps use `insecure_rand.randrange(m_consistency_check_ratio)`? Probably orders of magnitude faster...

MarcoFalke replied:
See previous discussion: https://github.com/bitcoin/bitcoin/pull/20233#discussion_r672913835

jnewbery replied:
I think I was mistaken about this causing problems for the unit tests. I'll try again with insecure_rand tomorrow.

jnewbery replied:
Done.

mzumsande commented on src/test/addrman_tests.cpp:
```diff
@@ -34,7 +34,7 @@ class CAddrManTest : public CAddrMan
     void MakeDeterministic()
     {
-        nKey.SetNull();
+        nKey = uint256{1};
```
The same `MakeDeterministic()` also exists in `net_tests.cpp` and should be changed there too to prevent `net_tests/caddrdb_read` from failing (though I'm not sure why these tests are there, looks to me like they should be integrated into `addrman_tests.cpp` or removed if redundant).

jnewbery replied:
Oh good spot! There's also a deterministic addrman in the fuzzers.

I've added a `deterministic` argument to the `CAddrMan` ctor (similar to `TxRequestTracker`) to consolidate these into one place, so that the tests can just ask for a deterministic addrman rather than constructing and addrman and then updating its internals.

jnewbery committed:
[addrman] Add deterministic argument to CAddrMan ctor

Removes the need for tests to update nKey and insecure_rand after constructing
a CAddrMan.

jnewbery commented:
Thanks for the recent review comments @sipa @MarcoFalke @mzumsande. All comments should now be addressed.

MarcoFalke commented on src/addrman.cpp:
```diff
@@ -459,11 +459,14 @@ CAddrInfo CAddrMan::Select_(bool newOnly)
 
+    // Run consistency checks 1 in m_consistency_check_ratio times if enabled
+    if (m_consistency_check_ratio == 0) return 0;
+    if (insecure_rand.randrange(m_consistency_check_ratio) >= 1) return 0;
```
This is still true, right?

> This has the unfortunate effect that the checks are run on the same operations for every run of the unit tests. Having them picked at random means that eventually they're run at every operation in the unit tests.

From https://github.com/bitcoin/bitcoin/pull/20233#discussion_r672913835

jnewbery replied:
Yes, I think this is true, but I'm not sure how big a problem it is. Before this PR we weren't running the consistency tests at all in the unit tests.

KOOLMOLL replied:
Test

vasild reviewed (approved):
ACK 6818d5e82f0a2ce5a7092c5a2b9d0760eca60866

> @vasild how many samples did you take?

Two from the PR and two with `master` and the variance seemed low. However when I run this again five times I see the timing wildly varies between 5min and 8min for both PR and `master`. Btw that is a debug build, not meant to be a performance benchmark but rather how long would it take on the dev's machines and on CI (which usually run debug builds).

jnewbery commented on src/addrman.h:
```diff
@@ -486,9 +489,15 @@ class CAddrMan
-    CAddrMan()
+    CAddrMan(bool deterministic = false,
+             int32_t consistency_check_ratio = DEFAULT_ADDRMAN_CONSISTENCY_CHECKS)
+        : m_consistency_check_ratio(consistency_check_ratio)
```
I've moved the `insecure_rand` to the initializer list as suggested.

I'm not going to remove the `nKey` assignment in this PR. I'd like to remove the `Clear()` method from `CAddrMan` in a future PR. The only place it's used in the product code is if peers.dat deserialization fails, we'll call `Clear()`. I think it's better just to instantiate a new `CAddrMan` object in that case, since it's definitely possible to imagine new data being added to `CAddrMan` and not added to the `Clear()` method. It wasn't previously possible to instantiate a new `CAddrMan` since it was a global initialized before `main()`, but now that its lifetime is managed to `context.node`, we can just reset the unique ptr to a new instance.

vasild replied:
Maybe also adding `explicit` would make sense.

jnewbery replied:
Made the ctor explicit.

jnewbery commented:
Thanks for the review @vasild! I've taken your suggestion of setting `insecure_rand` in the initializer list.

I've also removed the default args for the `CAddrMan` constructor, since default args are evil and can hide subtle bugs.

vasild commented on src/bench/addrman.cpp:
```diff
@@ -82,7 +82,7 @@ static void AddrManAdd(benchmark::Bench& bench)
 static void AddrManSelect(benchmark::Bench& bench)
 {
-    CAddrMan addrman;
+   CAddrMan addrman(/* deterministic */ false, /* consistency_checks */ 0);
```
nit: here and below - indentation of 3 spaces

jnewbery replied:
oops. Fixed.

vasild commented on src/test/fuzz/connman.cpp:
```diff
@@ -25,7 +25,7 @@ FUZZ_TARGET_INIT(connman, initialize_connman)
     FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
     SetMockTime(ConsumeTime(fuzzed_data_provider));
-    CAddrMan addrman;
+    CAddrMan addrman(/* deterministic */ false, /* consistency_checks */ 0);
```
In the previous incarnation of this PR which did not pass `consistency_checks` explicitly: `CAddrMan addrman(/* deterministic */ false)` it would have equaled to `DEFAULT_ADDRMAN_CONSISTENCY_CHECKS`.

In the latest incarnation (24515fee60) `0` is passed explicitly.

This contains the assumption that `DEFAULT_ADDRMAN_CONSISTENCY_CHECKS` is `0` which is true now, but could change in the future.

I think it is good to keep using `DEFAULT_ADDRMAN_CONSISTENCY_CHECKS` as default value instead of `0`. That is, either call `CAddrMan addrman(/* deterministic */ false, DEFAULT_ADDRMAN_CONSISTENCY_CHECKS)` in all places or restore the default second argument, which defaults to `DEFAULT_ADDRMAN_CONSISTENCY_CHECKS`.

jnewbery replied:
It's the other way around. The previous version implicitly assumed that the `DEFAULT_ADDRMAN_CONSISTENCY_CHECKS` was 0. This line is explicitly saying that we want an addrman without consistency checks.

If we used `DEFAULT_ADDRMAN_CONSISTENCY_CHECKS` explicitly in the call sites and that value changes in future (unlikely I think, since we don't want these debug checks enabled in production), then all call sites would have to be checked to make sure that the change in behavior is what is wanted. For example, we certainly wouldn't want to enable the consistency checks in benchmarks, since it'd totally skew the measurements.

vasild replied:
I see, makes sense, thanks for the elaboration!

vasild reviewed (approved):
ACK 0f959391190d973f09ea8ba5d4f34a87e2164040

vasild commented:
Before this PR, it was possible to run all unit and fuzz tests with checks enabled on every step (when compiled with `DEBUG_ADDRMAN`). One had to recompile, but changing the source code was not necessary.

With this PR, addrman unit tests are run with checks on every 100th step, other unit tests and fuzz tests have the checks disabled. To change that (to enable the checks or get them more often than once per 100), a modification of the source code is necessary on the following lines:

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/addrman_tests.cpp#L25

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/fuzz/addrman.cpp#L29

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/fuzz/connman.cpp#L28

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/fuzz/data_stream.cpp#L24

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/fuzz/deserialize.cpp#L202

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/net_tests.cpp#L38

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/net_tests.cpp#L122

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/net_tests.cpp#L139

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/net_tests.cpp#L153

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/net_tests.cpp#L169

https://github.com/bitcoin/bitcoin/blob/0f959391190d973f09ea8ba5d4f34a87e2164040/src/test/util/setup_common.cpp#L196

What about replacing those hardcoded `0`s (no checks) with a constant `DEFAULT_ADDRMAN_CONSISTENCY_CHECKS_IN_TESTS` that can be easily changed in order to enable the checks? Or maybe have two constants `..._IN_ADDRMAN_TESTS=100` and `..._IN_TESTS=0`.

Or replace the `0`s with something like `m_args.GetArg("-checkaddrman", 0)` so that the checks can be enabled without modifying the source code, e.g.:
```
./src/test/test_bitcoin --run_test="net_tests/*" -- -checkaddrman=5
```

Commit:
Add missing const to CAddrMan::Check_()

Also: Always compile the function signature to avoid similar issues in
the future.

jnewbery commented:
Rebased and taken the fix from #22601.

> What about replacing those hardcoded 0s (no checks) with a constant DEFAULT_ADDRMAN_CONSISTENCY_CHECKS_IN_TESTS that can be easily changed in order to enable the checks? Or maybe have two constants ..._IN_ADDRMAN_TESTS=100 and ..._IN_TESTS=0.

> Or replace the 0s with something like m_args.GetArg("-checkaddrman", 0) so that the checks can be enabled without modifying the source code.

@vasild - This can be done in a follow-up PR if desired. This PR has already gone through multiple rounds of review, so I'd like to settle on a final version of it now.

theStack commented:
Concept ACK

This approach seems to resolve a lot of potential headache caused by the fact that barely anyone ever builds with DEBUG_ADDRMAN (see https://github.com/bitcoin/bitcoin/pull/21940#issuecomment-890937517 for an example where this caused problems) and we don't have a CI instance running with it. The only slight drawback I could think of this solution is that the (released) bitcoind binary is a little bit larger compared to the current compile-time flag approach, containing code that the average user will never run. OTOH the fact that with this the tests can trigger regular addrman consistency checks with the default build outweights this by far :)

jonatack commented on src/addrman.h:
```diff
@@ -119,7 +122,7 @@ class CAddrInfo : public CAddress
  *      be observable by adversaries.
- *    * Several indexes are kept for high performance. Defining DEBUG_ADDRMAN will introduce frequent (and expensive)
+ *    * Several indexes are kept for high performance. Setting m_consistency_check_ratio will introduce (expensive)
  *      consistency checks for the entire data structure.
```
7883160 suggestion
```diff
- *    * Several indexes are kept for high performance. Setting m_consistency_check_ratio will introduce (expensive)
- *      consistency checks for the entire data structure.
+ *    * Several indexes are kept for high performance. Setting m_consistency_check_ratio with the -checkaddrman
+        config option will introduce (expensive) consistency checks for the entire data structure.
```

jnewbery replied:
done

jonatack commented on src/init.cpp:
```diff
@@ -501,6 +501,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-checklevel=<n>", strprintf("How thorough the block verification of -checkblocks is: %s (0-4, default: %u)", Join(CHECKLEVEL_DOC, ", "), DEFAULT_CHECKLEVEL), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-checkblockindex", strprintf("Do a consistency check for the block tree, chainstate, and other validation data structures occasionally. (default: %u, regtest: %u)", defaultChainParams->DefaultConsistencyChecks(), regtestChainParams->DefaultConsistencyChecks()), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-checkaddrman", strprintf("Do consistency checks on the addrman every <n> operations. Use 0 to disable. (default: %u)", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-checkmempool=<n>", strprintf("Run checks every <n> transactions (default: %u, regtest: %u)", defaultChainParams->DefaultConsistencyChecks(), regtestChainParams->DefaultConsistencyChecks()), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
7883160 suggestion
```diff
-    argsman.AddArg("-checkaddrman", strprintf("Do consistency checks on the addrman every <n> operations. Use 0 to disable. (default: %u)", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
-    argsman.AddArg("-checkmempool=<n>", strprintf("Run checks every <n> transactions (default: %u, regtest: %u)", defaultChainParams->DefaultConsistencyChecks(), regtestChainParams->DefaultConsistencyChecks()), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-checkaddrman=<n>", strprintf("Run addrman consistency checks every <n> operations. Use 0 to disable. (default: %u)", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-checkmempool=<n>", strprintf("Run mempool consistency checks every <n> transactions (default: %u, regtest: %u)", defaultChainParams->DefaultConsistencyChecks(), regtestChainParams->DefaultConsistencyChecks()), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```

jnewbery replied:
It's the same default for both mainnet and regtest (disabled).

jnewbery replied:
Done

jonatack commented on src/test/net_tests.cpp:
```diff
@@ -105,7 +101,6 @@ BOOST_AUTO_TEST_CASE(cnode_listen_port)
 BOOST_AUTO_TEST_CASE(caddrdb_read)
 {
     CAddrManUncorrupted addrmanUncorrupted;
-    addrmanUncorrupted.MakeDeterministic();
```
579584d Should the 4 CAddrMan ctors that follow pass deterministic `true` instead of `false`?

```diff
-    CAddrMan addrman1(/* deterministic */ false, /* consistency_checks */ 0);
+    CAddrMan addrman1(/* deterministic */ true, /* consistency_checks */ 0);
```

jnewbery replied:
I don't think so. They weren't deterministic before.

jonatack commented on src/bench/addrman.cpp:
```diff
@@ -112,10 +112,12 @@ static void AddrManGood(benchmark::Bench& bench)
-    for (auto& addrman : addrmans) {
-        FillAddrMan(addrman);
+    std::vector<std::unique_ptr<CAddrMan>> addrmans(addrman_count);
+    for (size_t i{0}; i < addrman_count; i++) {
```
579584d nit, prefer prefix iterator (idem in addrman_tests.cpp)
```suggestion
    for (size_t i{0}; i < addrman_count; ++i) {
```

jnewbery replied:
Fixed here. I don't think I've added any `i++` in addrman_tests.cpp.

jonatack commented on src/bench/addrman.cpp:
```diff
@@ -112,10 +112,12 @@ static void AddrManGood(benchmark::Bench& bench)
-        FillAddrMan(addrman);
+    std::vector<std::unique_ptr<CAddrMan>> addrmans(addrman_count);
+    for (size_t i{0}; i < addrman_count; i++) {
+        addrmans[i] = std::make_unique<CAddrMan>(/* deterministic */ false, /* consistency_checks */ 0);
```
579584d for benchmarks would it be preferable to pass deterministic `true`?

jnewbery replied:
Perhaps, but that's a change in behaviour. Maybe we can do it in a follow up PR.

jonatack commented on src/addrman.h:
```diff
@@ -691,6 +697,9 @@ class CAddrMan
     std::set<int> m_tried_collisions;
 
+    /** Perform consistency checks every m_consistency_check_ratio operations (if non-zero). */
+    const int32_t m_consistency_check_ratio;
```
7883160 negative values would be meaningless and value is clamped in range 0...1000000; could be unsigned

<details><summary>diff</summary><p>

```diff
--- a/src/addrman.h
+++ b/src/addrman.h
@@ -27,7 +27,7 @@ 
 /** Default for -checkaddrman */
 /** Default for -checkaddrman */
-static constexpr int32_t DEFAULT_ADDRMAN_CONSISTENCY_CHECKS{0};
+static constexpr uint32_t DEFAULT_ADDRMAN_CONSISTENCY_CHECKS{0};
 
@@ -495,7 +495,7 @@ public: 
-    explicit CAddrMan(bool deterministic, int32_t consistency_check_ratio)
+    explicit CAddrMan(bool deterministic, uint32_t consistency_check_ratio)

@@ -704,7 +704,7 @@ private:
     /** Perform consistency checks every m_consistency_check_ratio operations (if non-zero). */
-    const int32_t m_consistency_check_ratio;
+    const uint32_t m_consistency_check_ratio;

--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1165,7 +1165,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
     assert(!node.addrman);
-    auto check_addrman = std::clamp<int32_t>(args.GetArg("-checkaddrman", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), 0, 1000000);
+    auto check_addrman = std::clamp<uint32_t>(args.GetArg("-checkaddrman", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), 0, 1000000);
```
</p></details>

jnewbery replied:
Since we're already clamping the value, I don't think it matters.

jonatack reviewed (commented):
ACK 7883160288226ddc9a682e48647efef051a6286e code review, rebased to current master, and at each commit debug-built with `CPPFLAGS="-DDEBUG_ADDRMAN"` and ran net_tests, addrman_tests, and `./src/bench/bench_bitcoin -filter=AddrManGood`. A number of suggestions below for this PR or follow-ups. Happy to re-ACK if any are taken.

Noting here these follow-ups in the comments that look worthwhile:
- https://github.com/bitcoin/bitcoin/pull/20233#discussion_r672883161
- https://github.com/bitcoin/bitcoin/pull/20233#issuecomment-889813074
- https://github.com/bitcoin/bitcoin/pull/20233#discussion_r676730241

jonatack commented on src/init.cpp:
```diff
@@ -1159,7 +1160,8 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     assert(!node.addrman);
-    node.addrman = std::make_unique<CAddrMan>();
+    auto check_addrman = std::clamp<int32_t>(args.GetArg("-checkaddrman", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), 0, 1000000);
+    node.addrman = std::make_unique<CAddrMan>(/* determinstic */ false, /* consistency_checks */ check_addrman);
```
78831602
```bash
$ test/lint/lint-spelling.sh 
src/init.cpp:1164: determinstic ==> deterministic
```

jnewbery replied:
fixed

amitiuttarwar commented on src/test/fuzz/addrman.cpp:
```diff
@@ -26,7 +26,7 @@ FUZZ_TARGET_INIT(addrman, initialize_addrman)
     FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
     SetMockTime(ConsumeTime(fuzzed_data_provider));
-    CAddrMan addr_man(/* deterministic */ true);
+    CAddrMan addr_man(/* deterministic */ true, /* consistency_checks */ 0);
```
Could it make sense to enable the consistency checks here? 

I saw the conversation here https://github.com/bitcoin/bitcoin/pull/20233#issuecomment-755188580, where you said: 
> I think somewhere that they're definitely useful would be in fuzz testing, to ensure that there's no way to violate addrman's invariants.

So wondering why you landed on disabling.

jnewbery replied:
Added here and to the other fuzzers.

amitiuttarwar reviewed (commented):
light code review ACK 7883160288

I think its great that, with this PR, the addrman unit tests would run the consistency checks, and it's easier for developers to manually run or to potentially incorporate into the functional tests.

jnewbery commented on src/bench/addrman.cpp:
```diff
@@ -112,10 +112,12 @@ static void AddrManGood(benchmark::Bench& bench)
      * we want to do the same amount of work in every loop iteration. */
 
     bench.epochs(5).epochIterations(1);
+    size_t addrman_count = bench.epochs() * bench.epochIterations();
```
done

jnewbery commented on src/test/addrman_tests.cpp:
```diff
@@ -256,24 +256,27 @@ BOOST_AUTO_TEST_CASE(addrman_new_collisions)
     CNetAddr source = ResolveIP("252.2.2.2");
 
-    BOOST_CHECK_EQUAL(addrman.size(), 0U);
+    uint32_t addrs{0};
```
done!

jnewbery commented:
Rebased and addressed review comments from @jonatack and @amitiuttarwar

jnewbery commented:
> In 0d2d0de the commit message states "addrman_tests fail when consistency checks are enabled"...I was unable to reproduce this failure in the preceding commits. Is it still the case, should the message be updated, can the commit be dropped?

@jonatack - I think it's still true. Try enabling `DEBUG_ADDRMAN` and running the addrman tests here: https://github.com/jnewbery/bitcoin/tree/2021-08-fail-addrman-tests

The first commit fixes the build so it's possible to build with `DEBUG_ADDRMAN`. The second commit aborts if `Check()` fails, rather than just logging.

Without the _Make deterministic addrman use nKey = 1_ commit, then when we make addrman checks a runtime option and add the assert, then the tests would fail.

jonatack commented:
Thanks @jnewbery, I'll have a look (once the guix build finishes ⏰)

amitiuttarwar commented:
reACK 00fd0891ce

jonatack commented:
ACK 00fd0891ce799084fdd137fafc0522c27fbfb429 per `git range-diff 03826ae 7883160 00fd089` modulo the addrman fuzzer crash

<details><summary>reproduced the CI addrman fuzzer crash locally</summary><p>

```bash
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior serialize.h:205:93 in 
fuzz: ./addrman.h:754: void CAddrMan::Check() const: Assertion `false' failed.
==62677== ERROR: libFuzzer: deadly signal
    #0 0x55f7ef82d671 in __sanitizer_print_stack_trace (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x1000671)
    #1 0x55f7ef775008 in fuzzer::PrintStackTrace() fuzzer.o
    #2 0x55f7ef758ef3 in fuzzer::Fuzzer::CrashCallback() fuzzer.o
    #3 0x7fe5896ab13f  (/lib/x86_64-linux-gnu/libpthread.so.0+0x1413f)
    #4 0x7fe589317ce0 in __libc_signal_restore_set signal/../sysdeps/unix/sysv/linux/internal-signals.h:86:3
    #5 0x7fe589317ce0 in raise signal/../sysdeps/unix/sysv/linux/raise.c:48:3
    #6 0x7fe589301536 in abort stdlib/abort.c:79:7
    #7 0x7fe58930140e in __assert_fail_base assert/assert.c:92:3
    #8 0x7fe589310661 in __assert_fail assert/assert.c:101:3
    #9 0x55f7ef86ed17 in CAddrMan::Check() const ./addrman.h:754:13
    #10 0x55f7ef866d2c in CAddrMan::Select(bool) const ./addrman.h:601:9
    #11 0x55f7ef8638e1 in addrman_fuzz_target(Span<unsigned char const>) test/fuzz/addrman.cpp:306:26
    #12 0x55f7ef85b9d5 in void std::__invoke_impl<void, void (*&)(Span<unsigned char const>), Span<unsigned char const> >(std::__invoke_other, void (*&)(Span<unsigned char const>), Span<unsigned char const>&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14
    #13 0x55f7ef85b89d in std::enable_if<is_invocable_r_v<void, void (*&)(Span<unsigned char const>), Span<unsigned char const> >, void>::type std::__invoke_r<void, void (*&)(Span<unsigned char const>), Span<unsigned char const> >(void (*&)(Span<unsigned char const>), Span<unsigned char const>&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:110:2
    #14 0x55f7ef85b608 in std::_Function_handler<void (Span<unsigned char const>), void (*)(Span<unsigned char const>)>::_M_invoke(std::_Any_data const&, Span<unsigned char const>&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:291:9
    #15 0x55f7f078934d in std::function<void (Span<unsigned char const>)>::operator()(Span<unsigned char const>) const /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:622:14
    #16 0x55f7f0788ff8 in LLVMFuzzerTestOneInput test/fuzz/fuzz.cpp:91:5
    #17 0x55f7ef75a783 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) fuzzer.o
    #18 0x55f7ef759c8a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) fuzzer.o
    #19 0x55f7ef75c084 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) fuzzer.o
    #20 0x55f7ef75c299 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) fuzzer.o
    #21 0x55f7ef74bae7 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) fuzzer.o
    #22 0x55f7ef7757c2 in main (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0xf487c2)
    #23 0x7fe589302d09 in __libc_start_main csu/../csu/libc-start.c:308:16
    #24 0x55f7ef722969 in _start (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0xef5969)

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 0 ; base unit: 0000000000000000000000000000000000000000
0x0,0x0,0x0,0x0,0x0,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0xe1,0xdc,0xdc,0xdc,0xdc,0xdc,0xdc,0xdc,0x23,0x23,0x23,0xbb,0x1,0x0,0x0,0x23,0x27,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x40,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x0,0x0,0x0,0x0,0x1,0x0,0x0,0x0,0x0,0x0,0xf,0xda,0x3f,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x23,0x23,0x0,0x0,0x0,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0xff,0xff,0xff,0x1f,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0x23,0xf4,0xf4,0xf4,0xf4,0xf4,0xf4,0xf4,0xf4,0xf4,0xb4,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0x23,0x23,0x23,0x23,0x23,0x23,0xe1,0xdc,0xdc,0xdc,0xdc,0x32,0xa3,0x1,0x0,0x0,0x0,0x0,0x0,0x0,0x41,0x0,0xff,0x0,0x0,0xff,0x0,0x0,0x2b,
\x00\x00\x00\x00\x00#############\xe1\xdc\xdc\xdc\xdc\xdc\xdc\xdc###\xbb\x01\x00\x00#'#############@#################\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x0f\xda?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00##\x00\x00\x00########\xff\xff\xff\x1f####################\xf4\xf4\xf4\xf4\xf4\xf4\xf4\xf4\xf4\xb4\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff######\xe1\xdc\xdc\xdc\xdc2\xa3\x01\x00\x00\x00\x00\x00\x00A\x00\xff\x00\x00\xff\x00\x00+
artifact_prefix='./'; Test unit written to ./crash-d20c926b04d9e4e2657d97d3c5e77bde0e9a9daa
Base64: AAAAAAAjIyMjIyMjIyMjIyMj4dzc3Nzc3NwjIyO7AQAAIycjIyMjIyMjIyMjIyMjQCMjIyMjIyMjIyMjIyMjIyMjAAAAAAEAAAAAAA/aPwAAAAAAAAAAAAAAIyMAAAAjIyMjIyMjI////x8jIyMjIyMjIyMjIyMjIyMjIyMjI/T09PT09PT09LT///////////////////8jIyMjIyPh3Nzc3DKjAQAAAAAAAEEA/wAA/wAAKw==
```
</p></details>

Don't hesitate to reconsider taking the suggestion at https://github.com/bitcoin/bitcoin/pull/20233#discussion_r682768076.


> Without the _Make deterministic addrman use nKey = 1_ commit, then when we make addrman checks a runtime option and add the assert, then the tests would fail.

Got it. You are right--I dropped that commit and the addrman unit tests fail.

jnewbery commented:
I've disabled the consistency checks again in the addrman fuzz test. I think the problem is that the fuzzer will call `Clear()`, and subsequent calls to `Check()` may fail, since `Clear()` may leave the CAddrMan in a slightly inconsistent state.

I have a branch that removes `Clear()` at https://github.com/jnewbery/bitcoin/tree/2021-08-remove-addrman-clear, which I'll open after this is merged, and then re-enable consistency checks in the fuzzer.

jonatack commented:
ACK a4d78546b0858602c60c03fdf8b35ca666ab2e56 per `git diff 00fd089 a4d7854`, tested by adding logging similar to #22479 and running with `-checkaddrman=<n>` for various values 0/1/10/100 etc, tested the updated docs with `bitcoind -help-debug | grep -A2 "checkaddrman\|checkmempool"` and verified rebased on master that compiling with `CPPFLAGS="-DDEBUG_ADDRMAN"` no longer causes the build to error.

```diff
src/addrman.cpp
@@ -438,6 +438,7 @@ int CAddrMan::Check_() const
     // Run consistency checks 1 in m_consistency_check_ratio times if enabled
     if (m_consistency_check_ratio == 0) return 0;
     if (insecure_rand.randrange(m_consistency_check_ratio) >= 1) return 0;
+    LogPrintf("Check addrman\n");
```

```
$ ./src/bitcoind -help-debug | grep -A2 "checkaddrman\|checkmempool"
  -checkaddrman=<n>
       Run addrman consistency checks every <n> operations. Use 0 to disable.
       (default: 0)
--
  -checkmempool=<n>
       Run mempool consistency checks every <n> transactions. Use 0 to disable.
       (default: 0, regtest: 1)
```

mzumsande commented:
Code-review ACK a4d78546b0858602c60c03fdf8b35ca666ab2e56
Also did some testing with different values of `-checkaddrman` and it worked as expected.

theStack commented on src/test/addrman_tests.cpp:
```diff
@@ -862,20 +868,20 @@ BOOST_AUTO_TEST_CASE(addrman_noevict)
 
-    BOOST_CHECK(addrman.size() == 23);
-    BOOST_CHECK(addrman.SelectTriedCollision().ToString() == "250.1.1.19:0");
+    BOOST_CHECK(addrman.size() == 36);
```
```suggestion
    BOOST_CHECK_EQUAL(addrman.size(), 36);
```

theStack commented on src/test/addrman_tests.cpp:
```diff
@@ -885,17 +891,17 @@ BOOST_AUTO_TEST_CASE(addrman_noevict)
-    BOOST_CHECK(addrman.size() == 33);
+    BOOST_CHECK(!addrman.Add(CAddress(addr36, NODE_NONE), source));
+    addrman.Good(addr36);
+    BOOST_CHECK(addrman.size() == 59);
```
```suggestion
    BOOST_CHECK_EQUAL(addrman.size(), 59);
```

theStack reviewed (approved):
Code-review ACK a4d78546b0858602c60c03fdf8b35ca666ab2e56

Agree with vasild's idea of introducing constants for the `consistency_checks` CAddrMan ctor param in unit tests, which can be done in a follow-up: https://github.com/bitcoin/bitcoin/pull/20233#issuecomment-889813074

Also left two yocto-nits below regarding unit test assertion style, probably also best picked up in a follow-up.

fanquake commented:
The various nits/changes mentioned can be done in a followup. Either separately, or potentially as part of the `Clear()` branch.

mzumsande commented:
> I've disabled the consistency checks again in the addrman fuzz test. I think the problem is that the fuzzer will call `Clear()`, and subsequent calls to `Check()` may fail, since `Clear()` may leave the CAddrMan in a slightly inconsistent state.
> 
> I have a branch that removes `Clear()` at https://github.com/jnewbery/bitcoin/tree/2021-08-remove-addrman-clear, which I'll open after this is merged, and then re-enable consistency checks in the fuzzer.

Not sure if this is sufficient. Since #22493, we have a fuzz test that deserializes from random data, and `Unserialize()` doesn't seem to be designed to prevent the finer inconsistencies from occurring that `Check()` checks for. I verified that the fuzz inputs from #22503 and #22519 still crash when running on this PR with `consistency_check_ratio=1` (#22504 doesn't for some reason)
`
